### ISSUE TITLE
Een verplicht attribuut met stereotype 'Gestructuureerd datatype'  MAG NOOIT minOccurs=0 hebben in het XSD.

### ISSUE THREAD

**Issue by wilkoquak** [Imvertor/Imvertor-Maven#552]
In issue #393 is beschreven dat in de mapping van UML naar gml-xsd in sommige gevallen een propertyType een minOccurs=0 krijgt bij een verplicht attribuut waar je dus minOccurs=1 zou verwachten.  Dit is een vervelende bug die in de praktijk problemen oplevert omdat het xsd hierdoor bestanden goedkeurt waarbij verplichte gegevens ontbreken. 

Op lange termijn moet dit probleem opgelost worden door de UML naar XSD specificatie van de GML standaard beter te implementeren. Deze wordt beschreven in issue #393. Op korte termijn zit er een bug in de huidige mapping die ik er graag uit wil hebben. Daarvoor is dit nieuwe issue gemaakt. Hieronder beschrijf ik wat er moet gebeuren.

**ALS** een object een **verplicht** attribuut heeft met een type dat het stereotype \<\<Gestructureerd datatype\>\> heeft **DAN MOET** de de minOccurs in het xsd dezelfde waarde hebben als de minimum cardinaliteit in het UML (en dus niet 0).

**Comment by ArjanLoeffen** [comment]
Wanneer kan dit worden doorgevoerd? Het lijkt nog steeds de vorm van en discussie te hebben. Er is geen assignment gedaan en, geen label. Het issue (en de refereerde issues) hangen een beetje in de lucht. Graag uitsluitsel.

**Comment by wilkoquak** [comment]
Over dit issue is geen disucssie meer. Deze graag zo snel mogelijk oplossen. Liefst binnen twee weken want zo rond die tijd moet ik een nieuwe versie van IMGeluid oplevereren waar dit speelt

Met het fixen van dit issue is het gerelateerd issue wat mij betreft een enhancement geworden die geparkeerd mag worden.

Ik heb de issue niet aan iemand toegewezen omdat ik daar de rechten niet voor heb @ArjanLoeffen

**Comment by ArjanLoeffen** [comment]
Is dit nu niet opgelost?

**Comment by wilkoquak** [comment]
Nee, het werkt nog steeds niet. Ik heb zojuist met xsd van IMGeluid geprobeerd te maken en voor de elementen met het stereotype <<gestructureerd datatype>> wordt nog steeds een propertytpe met minOccurs=0 aangemaakt.

**Comment by ArjanLoeffen** [comment]
We moeten hier samen naar kijken, ik weet even niet waar ik naar moet zoeken.

**Comment by ArjanLoeffen** [comment]
Na overleg met @wilkoquak is duidelijk dat hier iets niet okay is. Het XML schema is wat deze situatie betreft niet strikt genoeg.
Neem het testgeval:

![image](https://github.com/user-attachments/assets/e555ef1f-ba4c-474a-b636-d30893bd6f21)

Dit levert nu het schema:

```xml
  <xs:element name="Objecttype1"
               type="Model:Objecttype1Type"
               substitutionGroup="gml:AbstractFeature"/>
   <xs:complexType name="Objecttype1Type">
      <xs:complexContent>
         <xs:extension base="gml:AbstractFeatureType">
            <xs:sequence>
               <xs:element name="att1" type="Model:GestructureerdDatatype1PropertyType"/>
            </xs:sequence>
         </xs:extension>
      </xs:complexContent>
   </xs:complexType>
   <xs:complexType name="Objecttype1PropertyType">
      <xs:sequence minOccurs="0">
         <xs:element ref="Model:Objecttype1"/>
      </xs:sequence>
      <xs:attributeGroup ref="gml:AssociationAttributeGroup"/>
      <xs:attributeGroup ref="gml:OwnershipAttributeGroup"/>
   </xs:complexType>
   <xs:element name="GestructureerdDatatype1" type="Model:GestructureerdDatatype1Type"/>
   <xs:complexType name="GestructureerdDatatype1Type">
      <xs:sequence>
         <xs:element name="data1" type="xs:string"/>
         <xs:element name="data2" type="xs:string"/>
      </xs:sequence>
   </xs:complexType>
   <xs:complexType name="GestructureerdDatatype1PropertyType">
      <xs:sequence minOccurs="0">
         <xs:element ref="Model:GestructureerdDatatype1"/>
      </xs:sequence>
   </xs:complexType>
``` 

Hiermee is het volgende toegestaan:

```xml
<Model:Objecttype1>
    <Model:att1/>
</Model:Objecttype1>
```

Dit moet dus zijn:

```xml
<xs:sequence> <!-- d.w.z. minOccurs=1 -->
     <xs:element ref="Model:GestructureerdDatatype1"/>
</xs:sequence>
```

Hiermee wordt het volgende afgedwongen:

```xml
<Model:Objecttype1>
    <Model:att1>
        <Model:GestructureerdDatatype1>
            <Model:data1>...</Model:data1>
            <Model:data2>...</Model:data2>
        </Model:GestructureerdDatatype1>
    </Model:att1>
</Model:Objecttype1>
```

NB dit geldt alleen voor gestruktureerde datatypen.
Uitgangspunt hierbij is dat de property [associateByReference ](https://armatiek.nl/doc/imvertor/imvertor-properties/1.0/#cli-assocbyreference) geen invloed heeft op deze aanpassing.

Het is duidelijk dat een verandering zoals deze mogelijk effect heeft op afnemers van ISOI 19136 schema's van Imvertor OS. @HanWelmer Voor jou even belangrijk om dit te reviewen.

**Comment by ArjanLoeffen** [comment]
Ik geef hier bij een aanvulling. Imvertor implementeert nu de volgende regel:

1. Is het een objecttype en is `associateByReference=yes`, dan wordt het objecttype niet opgenomen als mogelijke inhoud van het refererende element. Er wordt altijd gebruik gemaakt van xlink `href` relaties.
2. Is het een objecttype en is` associateByReference=no`, dan wordt het objecttype opgenomen als mogelijke inhoud van het refererende element. Men kan ook xlink `href` gebruiken.
3. Is het een keuze tussen relaties, dan wordt die keuze het objecttype opgenomen als mogelijke inhoud van het refererende element. Men kan ook xlink `href` gebruiken.
4. In alle andere gevallen wordt aangenomen dat het element zelf in de XML op de plek waar deze verwijzing plaatsvindt moet worden opgenomen. Xlink `href` is niet van toepassing.

Die laatste is het geval zoals in eerdere commentaar beschreven.

**Comment by ArjanLoeffen** [comment]
Volgens mij moet het nu goed gaan. Graag testen.

**Comment by wilkoquak** [comment]
Gedaan en het werkt. Dank!
